File: langs/zig/src/spec.zig
//! Clide specification data structures

const std = @import("std");

pub const Type = enum {
    int,
    bool,
    str,
    path,

    pub fn fromStr(s: []const u8) !Type {
        if (std.mem.eql(u8, s, "INT")) return .int;
        if (std.mem.eql(u8, s, "BOOL")) return .bool;
        if (std.mem.eql(u8, s, "STR")) return .str;
        if (std.mem.eql(u8, s, "PATH")) return .path;
        return error.UnknownType;
    }

    pub fn toString(self: Type) []const u8 {
        return switch (self) {
            .int => "INT",
            .bool => "BOOL",
            .str => "STR",
            .path => "PATH",
        };
    }
};

pub const Atom = union(enum) {
    lit: []const u8,
    opt_bool: OptBool,
    opt_val: OptVal,
    pos: Pos,

    pub const OptBool = struct {
        long: ?[]const u8,
        short: ?[]const u8,
    };

    pub const OptVal = struct {
        long: ?[]const u8,
        short: ?[]const u8,
        ty: Type,
        default: ?[]const u8,
        allow_repeat: bool,
    };

    pub const Pos = struct {
        name: []const u8,
        ty: Type,
        default: ?[]const u8,
    };
};

pub const Group = union(enum) {
    single: Atom,
    alt: []const Atom,

    pub fn atoms(self: *const Group, allocator: std.mem.Allocator) ![]const Atom {
        switch (self.*) {
            .single => |a| {
                const result = try allocator.alloc(Atom, 1);
                result[0] = a;
                return result;
            },
            .alt => |atoms| return atoms,
        }
    }
};

pub const Item = union(enum) {
    required: Group,
    optional: Group,
};

pub const Command = struct {
    name: []const u8,
    items: []const Item,
};

pub const Spec = struct {
    prog: []const u8,
    commands: []const Command,
};

File: langs/rust/src/spec.rs
//! Clide specification data structures

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Int,
    Bool,
    Str,
    Path,
}

impl Type {
    pub fn from_str(s: &str) -> Result<Self, String> {
        match s {
            "INT" => Ok(Type::Int),
            "BOOL" => Ok(Type::Bool),
            "STR" => Ok(Type::Str),
            "PATH" => Ok(Type::Path),
            _ => Err(format!("Unknown type: {}", s)),
        }
    }

    pub fn to_string(&self) -> &'static str {
        match self {
            Type::Int => "INT",
            Type::Bool => "BOOL",
            Type::Str => "STR",
            Type::Path => "PATH",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Atom {
    Lit(String),
    OptBool {
        long: Option<String>,
        short: Option<String>,
    },
    OptVal {
        long: Option<String>,
        short: Option<String>,
        ty: Type,
        default: Option<String>,
        allow_repeat: bool,
    },
    Pos {
        name: String,
        ty: Type,
        default: Option<String>,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Group {
    Single(Atom),
    Alt(Vec<Atom>),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Item {
    Required(Group),
    Optional(Group),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Command {
    pub name: String,
    pub items: Vec<Item>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Spec {
    pub prog: String,
    pub commands: Vec<Command>,
}

impl Group {
    pub fn atoms(&self) -> Vec<&Atom> {
        match self {
            Group::Single(a) => vec![a],
            Group::Alt(atoms) => atoms.iter().collect(),
        }
    }
}
